components/TARGET_PSA/services/crypto/COMPONENT_PSA_SRV_IPC/psa_crypto_spm.c

```diff
@@ -23,56 +23,96 @@
 
 #include <stdlib.h>
 #include <string.h>
+#include <stdbool.h>
 
 #include "psa_manifest/sid.h"
 #include "psa/client.h"
-#include "crypto.h"
+#include "psa/crypto.h"
 #include "crypto_platform_spe.h"
-
 #include "mbed_assert.h"
 
 #define MINOR_VER 1
 #define CLIENT_PSA_KEY_ID_SIZE_IN_BYTES 4
 
 MBED_STATIC_ASSERT(sizeof(psa_key_id_t) == CLIENT_PSA_KEY_ID_SIZE_IN_BYTES, "Unexpected psa_key_id_t size");
 
-psa_status_t psa_crypto_init(void)
+/****************************************************************/
+/* INTERNAL HELPER FUNCTIONS */
+/****************************************************************/
+static psa_status_t ipc_connect(uint32_t sid, psa_handle_t *handle)
 {
-    psa_status_t status;
-    psa_handle_t handle = PSA_NULL_HANDLE;
-
-    handle = psa_connect(PSA_CRYPTO_INIT_ID, MINOR_VER);
-    if (handle <= 0) {
+    *handle = psa_connect(sid, MINOR_VER);
+    if (*handle <= PSA_NULL_HANDLE) {
         return (PSA_ERROR_COMMUNICATION_FAILURE);
     }
+    return (PSA_SUCCESS);
+}
+
+static inline void ipc_close(psa_handle_t *handle)
+{
+    psa_close(*handle);
+    *handle = PSA_NULL_HANDLE;
+}
+
+static psa_status_t ipc_call(psa_handle_t *handle, psa_invec *in_vec, size_t in_vec_size,
+                             psa_outvec *out_vec, size_t out_vec_size, bool close)
+{
+    if (*handle <= PSA_NULL_HANDLE) {
+        return (PSA_ERROR_BAD_STATE);
+    }
 
-    status = psa_call(handle, NULL, 0, NULL, 0);
-    psa_close(handle);
+    psa_status_t status = psa_call(*handle, in_vec, in_vec_size, out_vec, out_vec_size);
+    if (close) {
+        ipc_close(handle);
+    }
+    return (status);
+}
 
+static psa_status_t ipc_oneshot(uint32_t sid, psa_invec *in_vec, size_t in_vec_size,
+                                psa_outvec *out_vec, size_t out_vec_size)
+{
+    psa_handle_t handle = PSA_NULL_HANDLE;
+    psa_status_t status = ipc_connect(sid, &handle);
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+    status = ipc_call(&handle, in_vec, in_vec_size, out_vec, out_vec_size, true);
     return (status);
 }
 
 /****************************************************************/
-/* PSA_MAC */
+/* MODULE SETUP/TEARDOWN */
 /****************************************************************/
+psa_status_t psa_crypto_init(void)
+{
+    psa_status_t status = ipc_oneshot(PSA_CRYPTO_INIT_ID, NULL, 0, NULL, 0);
+    return (status);
+}
 
-psa_status_t psa_mac_abort(psa_mac_operation_t *operation)
+void mbedtls_psa_crypto_free(void)
 {
-    psa_status_t status;
-    psa_crypto_ipc_t psa_crypto_ipc = { 0, 0, 0 };
+    //TODO: add retry mechanism to make sure resources were deallocated.
```

Comment:
We should not leave TODO in the code, create a tracking issue rather

Reply:
I removed the comment as it's not relevant anymore.